databricks_v0.3.7 crashed

Hi there,

Thank you for opening an issue. Please note that we try to keep the Databricks Provider issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Configuration
```hcl
resource "databricks_secret_scope" "terraform" {
    name                     = "application"
    initial_manage_principal = "users"
}

resource "databricks_secret" "service_principal_key" {
    key          = "service_principal_key"
    string_value = "${var.ARM_CLIENT_SECRET}"
    scope        = databricks_secret_scope.terraform.name
}

data "azurerm_client_config" "current" {
}

resource "azurerm_storage_account" "this" {
  name                     = "${var.prefix}datalake"
  resource_group_name      = var.resource_group_name
  location                 = var.resource_group_location
  account_tier             = "Standard"
  account_replication_type = "GRS"
  account_kind             = "StorageV2"
  is_hns_enabled           = true
}

resource "azurerm_role_assignment" "this" {
  scope                = azurerm_storage_account.this.id
  role_definition_name = "Storage Blob Data Contributor"
  principal_id         = data.azurerm_client_config.current.object_id
}

resource "azurerm_storage_container" "this" {
  name                  = "marketing"
  storage_account_name  = azurerm_storage_account.this.name
  container_access_type = "private"
}

resource "databricks_azure_adls_gen2_mount" "marketing" {
    container_name         = azurerm_storage_container.this.name
    storage_account_name   = azurerm_storage_account.this.name
    mount_name             = "marketing"
    tenant_id              = data.azurerm_client_config.current.tenant_id
    client_id              = data.azurerm_client_config.current.client_id
    client_secret_scope    = databricks_secret_scope.terraform.name
    client_secret_key      = databricks_secret.service_principal_key.key
    initialize_file_system = true
}
```

### Expected Behavior
What should have happened?
Storage databricks_azure_adls_gen2_mount should be mounted

### Actual Behavior
What actually happened?
Stack trace from the terraform-provider-databricks_v0.3.7 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0xea84de]

goroutine 37 [running]:
github.com/databrickslabs/terraform-provider-databricks/storage.AzureADLSGen2Mount.Config(0xc00062e3c0, 0x17, 0xc00062e408, 0x12, 0x0, 0x0, 0xc00057c180, 0x24, 0xc00057c1b0, 0x24, ...)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/storage/adls_gen2_mount.go:30 +0x3e
github.com/databrickslabs/terraform-provider-databricks/storage.MountPoint.Mount(0x131da80, 0xc0002cf800, 0xc000722978, 0x14, 0xc00030a5fc, 0x4, 0x1328840, 0xc000326380, 0xc00027e3c0, 0xc000326380, ...)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/storage/mounts.go:67 +0x59
github.com/databrickslabs/terraform-provider-databricks/storage.mountCreate.func1(0x1331a70, 0xc0006517a0, 0xc000363d00, 0x1115b00, 0xc00027e3c0, 0xc00065e7b0, 0x1331a70, 0xc0006517a0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/storage/mounts.go:207 +0x2da
github.com/databrickslabs/terraform-provider-databricks/common.addContextToStage.func1(0x1331a38, 0xc000631620, 0xc000363d00, 0x1115b00, 0xc00027e3c0, 0xc00065e7a0, 0xa9bb4a, 0xc000526ae0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/common/context.go:43 +0xf6
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc0005e6fc0, 0x13319c8, 0xc000323c40, 0xc000363d00, 0x1115b00, 0xc00027e3c0, 0x0, 0x0, 0x0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:330 +0x17f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0005e6fc0, 0x13319c8, 0xc000323c40, 0xc0006450a0, 0xc000526ae0, 0x1115b00, 0xc00027e3c0, 0x0, 0x0, 0x0, ...)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:456 +0x67b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00000cb40, 0x13319c8, 0xc000323c40, 0xc000335180, 0xc000323c40, 0x10dce80, 0xc000650800)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc000690dc0, 0x1331a70, 0xc000323c40, 0xc000644e70, 0xc000690dc0, 0xc000650810, 0xc000639ba0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x10dce80, 0xc000690dc0, 0x1331a70, 0xc000650810, 0xc000631020, 0x0, 0x1331a70, 0xc000650810, 0xc000654300, 0x2c8)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002fba40, 0x133ba98, 0xc000549080, 0xc0000bcea0, 0xc000302f00, 0x1942360, 0x0, 0x0, 0x0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/google.golang.org/grpc/server.go:1292 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0002fba40, 0x133ba98, 0xc000549080, 0xc0000bcea0, 0x0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/google.golang.org/grpc/server.go:1617 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000034580, 0xc0002fba40, 0x133ba98, 0xc000549080, 0xc0000bcea0)
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/google.golang.org/grpc/server.go:940 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/google.golang.org/grpc/server.go:938 +0x1fd

Error: The terraform-provider-databricks_v0.3.7 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`
- export TF_VAR_ARM_CLIENT_SECRET = "secret_value"
- terraform init
- terraform plan
- terraform apply
### Terraform and provider versions

Please paste the output of `terraform version`. If version of `databricks` provider is not the latest (https://github.com/databrickslabs/terraform-provider-databricks/releases), please make sure to use the latest one.

❯ terraform --version
Terraform v1.0.7
on linux_amd64
+ provider registry.terraform.io/databrickslabs/databricks v0.3.7
+ provider registry.terraform.io/hashicorp/azurerm v2.76.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/time v0.7.2

### Debug Output
Please add turn on logging, e.g. `TF_LOG=DEBUG terraform apply` and run command again, paste it to gist & provide the link to gist. If you're still willing to paste in log output, make sure you provide only relevant log lines with requests.

It would make it more readable, if you pipe the log through `| grep databricks | sed -E 's/^.* plugin[^:]+: (.*)$/\1/'`, e.g.:

```
TF_LOG=DEBUG terraform plan 2>&1 | grep databricks | sed -E 's/^.* plugin[^:]+: (.*)$/\1/'
```
https://gist.github.com/davkohli/18a17a9567170f505bca6b7fd6de2921
<script src="https://gist.github.com/davkohli/18a17a9567170f505bca6b7fd6de2921.js"></script>

If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Important Factoids
Are there anything atypical about your accounts that we should know?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 * Added support for Azure MSI authentication ([#743](https://github.com/databrickslabs/terraform-provider-databricks/pull/743))
 * Already deleted `databricks_token` don't fail the apply ([#808](https://github.com/databrickslabs/terraform-provider-databricks/pull/808))
 * Default `terraform-mount` clusters created for mounting for `databricks_aws_s3_mount`, `databricks_azure_adls_gen1_mount`, `databricks_azure_adls_gen2_mount`, and `databricks_azure_blob_mount` have now `spark.scheduler.mode` as `FIFO` ([#828](https://github.com/databrickslabs/terraform-provider-databricks/pull/828))
+* Fixed crash when using non-Azure authentication to mount Azure resources ([#831](https://github.com/databrickslabs/terraform-provider-databricks/issues/831))
 * Multiple documentation improvements
 
 **Deprecations**

--- a/common/azure_auth.go
+++ b/common/azure_auth.go
@@ -156,12 +156,7 @@ func (aa *DatabricksClient) simpleAADRequestVisitor(
 	ctx context.Context,
 	authorizerFactory func(resource string) (autorest.Authorizer, error),
 	visitors ...func(r *http.Request, ma autorest.Authorizer) error) (func(r *http.Request) error, error) {
-	azureEnvironment, err := aa.getAzureEnvironment()
-	if err != nil {
-		return nil, fmt.Errorf("cannot get azure environment: %w", err)
-	}
-	aa.AzureEnvironment = &azureEnvironment
-	managementAuthorizer, err := authorizerFactory(azureEnvironment.ServiceManagementEndpoint)
+	managementAuthorizer, err := authorizerFactory(aa.AzureEnvironment.ServiceManagementEndpoint)
 	if err != nil {
 		return nil, fmt.Errorf("cannot authorize management: %w", err)
 	}
@@ -203,12 +198,7 @@ func (aa *DatabricksClient) acquirePAT(
 	if aa.temporaryPat != nil {
 		return aa.temporaryPat, nil
 	}
-	azureEnvironment, err := aa.getAzureEnvironment()
-	if err != nil {
-		return nil, fmt.Errorf("cannot get azure environment: %w", err)
-	}
-	aa.AzureEnvironment = &azureEnvironment
-	management, err := factory(azureEnvironment.ServiceManagementEndpoint)
+	management, err := factory(aa.AzureEnvironment.ServiceManagementEndpoint)
 	if err != nil {
 		return nil, err
 	}
@@ -274,13 +264,9 @@ func (aa *DatabricksClient) ensureWorkspaceURL(ctx context.Context,
 	if resourceID == "" {
 		return fmt.Errorf("somehow resource id is not set")
 	}
-	env, err := aa.getAzureEnvironment()
-	if err != nil {
-		return maybeExtendAuthzError(err)
-	}
 	log.Println("[DEBUG] Getting Workspace ID via management token.")
 	// All azure endpoints typically end with a trailing slash removing it because resourceID starts with slash
-	managementResourceURL := strings.TrimSuffix(env.ResourceManagerEndpoint, "/") + resourceID
+	managementResourceURL := strings.TrimSuffix(aa.AzureEnvironment.ResourceManagerEndpoint, "/") + resourceID
 	var workspace azureDatabricksWorkspace
 	resp, err := aa.genericQuery(ctx, http.MethodGet,
 		managementResourceURL,

--- a/common/client.go
+++ b/common/client.go
@@ -203,6 +203,14 @@ func (c *DatabricksClient) Configure(attrsUsed ...string) error {
 	if c.DebugTruncateBytes == 0 {
 		c.DebugTruncateBytes = DefaultTruncateBytes
 	}
+	// AzureEnvironment could be used in the different contexts, not only for Auzre Authentication
+	// lack of this lead to crash (see issue #831)
+	azureEnvironment, err := c.getAzureEnvironment()
+	if err != nil {
+		return fmt.Errorf("cannot get azure environment: %w", err)
+	}
+	c.AzureEnvironment = &azureEnvironment
+
 	return nil
 }
 

